src/v/cluster/archival_metadata_stm.cc

```diff
@@ -512,7 +512,8 @@ ss::future<> archival_metadata_stm::apply(model::record_batch b) {
         _insync_offset = b.last_offset();
         co_return;
     }
-
+    // Lock asynchronous manifest serialization
+    auto units = co_await _manifest->get_serialization_units(_download_as);
```

Comment:
This is the only place where the manifest is updated, the `ntp_archiver` gets a constant reference. So it's the only place where we need to protect it. The accessors are not futurized so it's also doesn't make much sense to protect them. The only dangerous operation is serialization since it's the only accessor which is asynchronous and  it can "see" inconsistent state.

Reply:
And this `_insync_offset` doesn't belong to manifest. The one that belong is changed on line 543.

Reply:
Ah, indeed! Missed that this was a different `_insync_offset`.

Agreed about not futurizing the accessors.

Looking around a bit at other potentially risky spots, looks like `get_insync_offset()` is called from `handle_eviction()`. It looks like this is serialized by Raft's `apply()`, and therefore also shouldn't overlap with mutations.

Still, this seems to rely a bit on existing code and will be easy to break unintentionally. Maybe it's just an expectation of our codebase at large, but maybe we should still call out that the accessors (`get_insync_offset(), etc) shouldn't be called with scheduling points in between calls. Or we could merge some of the common accessors so there's an atomic unit to work with (e.g. getting both the insync offset and last offset in one call).